### PR TITLE
[CP] Add sending mail for hosts while adding a comment

### PR DESCRIPTION
**Pull request by illusionOfParadise**

### CODE REVIEW

File: src/CitizenProject/CitizenProjectCommentCreationCommandHandler.php [hunk 1]
```diff
@@ -25,5 +32,24 @@ public function handle(CitizenProjectCommentCommand $command): void
...
+    }
+
+    private function sendMessageToMembers(CitizenProjectComment $comment, CitizenProject $citizenProject): void
+    {
+        foreach ($this->getOptinCitizenProjectFollowersChunks($citizenProject) as $chunk) {
+            $this->mailer->sendMessage(CitizenProjectCommentMessage::create($chunk, $comment));
+        }
+    }
+
+    private function getOptinCitizenProjectFollowersChunks(CitizenProject $committee)
```
Comment (Remg) [inline src/CitizenProject/CitizenProjectCommentCreationCommandHandler.php]: missing return type hint

File: src/Form/CitizenProjectCommentCommandType.php [hunk 2]
```diff
@@ -3,18 +3,45 @@
...
+        $this->authorizationChecker = $authorizationChecker;
+    }
+
     public function buildForm(FormBuilderInterface $builder, array $options)
     {
         $builder
             ->add('content', TextType::class)
         ;
+
+        $builder->addEventListener(FormEvents::PRE_SET_DATA, function (FormEvent $event) {
```
Comment (HeahDude) [inline src/Form/CitizenProjectCommentCommandType.php]: If this form is used as root form please remove this listener, `$data->getCitizenProject()` might break if there is no initial data.

Please add something like:
```php
$data = $builder->getData();

if ($data instanceof CitizenProjectCommentCommand) {
    $builder
        ->add('sendMail', CheckboxType::class, [
            'required' => false,
            'label' => 'Envoyer aussi par e-mail',
            'attr' => ['class' => 'form__checkbox form__checkbox--large'],
        ])
    ;
}
```

File: tests/Controller/EnMarche/CitizenProjectControllerTest.php [hunk 3]
```diff
@@ -115,6 +116,44 @@ public function testFollowerCanAddCommentToCitizenProject(): void
...
     }
 
+    /**
+     * @depends testFollowerCanSeeACitizenProject
+     */
+    public function testFollowerCanNotSendCommentToCitizenProjectInMail(): void
+    {
+        $this->authenticateAsAdherent($this->client, 'carl999@example.fr', 'secret!12345');
+        $this->client->request(Request::METHOD_GET, '/projets-citoyens/le-projet-citoyen-a-paris-8/discussions');
+        $this->assertResponseStatusCode(Response::HTTP_OK, $this->client->getResponse());
```
Comment (HeahDude) [inline tests/Controller/EnMarche/CitizenProjectControllerTest.php]: CS: please add an empty line before assertions.

File: tests/Controller/EnMarche/CitizenProjectControllerTest.php [hunk 4]
```diff
@@ -115,6 +116,44 @@ public function testFollowerCanAddCommentToCitizenProject(): void
...
+    /**
+     * @depends testAdministratorCanSeeACitizenProject
+     */
+    public function testAdministratorCanAddCommentToCitizenProjectWithSendingMail(): void
+    {
+        $this->authenticateAsAdherent($this->client, 'jacques.picard@en-marche.fr', 'changeme1337');
+        $this->client->request(Request::METHOD_GET, '/projets-citoyens/le-projet-citoyen-a-paris-8/discussions');
+        $this->assertResponseStatusCode(Response::HTTP_OK, $this->client->getResponse());
+        $this->assertCount(1, $this->client->getCrawler()->filter('label:contains("Envoyer aussi par e-mail")'));
+        $this->client->submit(
```
Comment (HeahDude) [inline tests/Controller/EnMarche/CitizenProjectControllerTest.php]: Please add an empty line after assertions.

Comment (illusionOfParadise) [reply tests/Controller/EnMarche/CitizenProjectControllerTest.php]: Done

File: tests/Controller/EnMarche/CitizenProjectControllerTest.php (hunk 3 above)
Comment (illusionOfParadise) [reply tests/Controller/EnMarche/CitizenProjectControllerTest.php]: Done

File: src/Form/CitizenProjectCommentCommandType.php (hunk 2 above)
Comment (illusionOfParadise) [reply src/Form/CitizenProjectCommentCommandType.php]: Done

Comment (HeahDude) [reply src/Form/CitizenProjectCommentCommandType.php]: You did not remove the listener.

File: app/config/services/system.xml [hunk 5]
```diff
@@ -267,6 +267,10 @@
             <tag name="form.type" />
         </service>
 
+        <service id="AppBundle\Form\CitizenProjectCommentCommandType" autoconfigure="true" autowire="true" >
+            <tag name="form.type" />
```
Comment (HeahDude) [inline app/config/services/system.xml]: The service is auto configured.

File: app/config/services/system.xml [hunk 6]
```diff
@@ -267,6 +267,8 @@
             <tag name="form.type" />
         </service>
 
+        <service autoconfigure="true" autowire="true" id="AppBundle\Form\CitizenProjectCommentCommandType"/>
```
Comment (HeahDude) [inline app/config/services/system.xml]: Should be moved in `services.xml`.

File: src/CitizenProject/CitizenProjectCommentCreatedEvent.php [hunk 7]
```diff
@@ -0,0 +1,23 @@
+<?php
+
+namespace AppBundle\CitizenProject;
+
+use AppBundle\Entity\CitizenProject;
+use AppBundle\Entity\CitizenProjectComment;
+
+class CitizenProjectCommentCreatedEvent extends CitizenProjectEvent
```
Comment (HeahDude) [inline src/CitizenProject/CitizenProjectCommentCreatedEvent.php]: `CitizenProjectCommentEvent`

File: src/CitizenProject/CitizenProjectCommentCreatedEvent.php [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    private $newComment;
+
+    public function __construct(CitizenProject $citizenProject, CitizenProjectComment $newComment)
+    {
+        parent::__construct($citizenProject);
+
+        $this->newComment = $newComment;
+    }
+
+    public function getNewComment(): CitizenProjectComment
```
Comment (HeahDude) [inline src/CitizenProject/CitizenProjectCommentCreatedEvent.php]: `getComment`

File: src/CitizenProject/CitizenProjectCommentCreatedEvent.php [hunk 9]
```diff
@@ -0,0 +1,23 @@
+<?php
+
+namespace AppBundle\CitizenProject;
+
+use AppBundle\Entity\CitizenProject;
+use AppBundle\Entity\CitizenProjectComment;
+
+class CitizenProjectCommentCreatedEvent extends CitizenProjectEvent
+{
+    private $newComment;
```
Comment (HeahDude) [inline src/CitizenProject/CitizenProjectCommentCreatedEvent.php]: `$comment`

File: src/CitizenProject/CitizenProjectCommentCreationCommandHandler.php [hunk 10]
```diff
@@ -25,5 +32,12 @@ public function handle(CitizenProjectCommentCommand $command): void
 
         $this->manager->persist($comment);
         $this->manager->flush();
+
+        if ($command->isSendMail()) {
```
Comment (HeahDude) [inline src/CitizenProject/CitizenProjectCommentCreationCommandHandler.php]: This if should be in the listener.

File: src/CitizenProject/CitizenProjectMessageNotifier.php [hunk 11]
```diff
@@ -65,6 +66,13 @@ public function onCitizenProjectFollowerAdded(CitizenProjectFollowerAddedEvent $
         ));
     }
 
+    public function onCitizenProjectCommentCreated(CitizenProjectCommentCreatedEvent $commentCreatedEvent): void
```
Comment (HeahDude) [inline src/CitizenProject/CitizenProjectMessageNotifier.php]: `onCitizenProjectCommentCreated` => `sendCommentCreatedEmail`

File: src/Form/CitizenProjectCommentCommandType.php (hunk 2 above)
Comment (HeahDude) [reply src/Form/CitizenProjectCommentCommandType.php]: You still have to remove this listener.

File: src/Mailer/Message/CitizenProjectCommentMessage.php [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+            'Message de votre porteur de projet',
+            [
+                'citizen_project_host_firstname' => self::escape($comment->getAuthor()->getFirstName()),
+                'citizen_project_host_message' => $comment->getContent(),
+            ],
+            [],
+            $comment->getAuthor()->getEmailAddress()
+        );
+
+        $sender = $comment->getAuthor()->getFirstName().', porteur de votre projet citoyen';
```
Comment (HeahDude) [inline src/Mailer/Message/CitizenProjectCommentMessage.php]: Please add a var for `$comment->getAuthor()->getFirstName()` which is also used above.

Comment (HeahDude) [inline src/Mailer/Message/CitizenProjectCommentMessage.php]: Please add a var for `$comment->getAuthor()->getFirstName()` which is also used above.

File: tests/CitizenProject/CitizenProjectMessageNotifierTest.php [hunk 13]
```diff
@@ -118,6 +120,25 @@ public function testSendAdminitratorNotificationWhenFollowerAddedWithAdministrat
...
+        $producer = $this->createMock(ProducerInterface::class);
+        $mailer = $this->createMock(MailerService::class);
+        $member = $this->createAdministrator(LoadAdherentData::ADHERENT_2_UUID);
+        $comment = $this->createComment();
+        $manager = $this->createManager(null, $member);
+        $citizenProject = $this->createCitizenProject(LoadCitizenProjectData::CITIZEN_PROJECT_1_UUID, 'Paris 8e');
+        $committeeManager = $this->createMock(CommitteeManager::class);
+        $router = $this->createMock(RouterInterface::class);
+
+        $manager->expects($this->once())->method('getCitizenProjectMembers')->willReturn(new AdherentCollection());
```
Comment (HeahDude) [inline tests/CitizenProject/CitizenProjectMessageNotifierTest.php]: We usually put each call on a new line for mocks.

File: src/Form/CitizenProjectCommentCommandType.php (hunk 2 above)
Comment (illusionOfParadise) [reply src/Form/CitizenProjectCommentCommandType.php]: We need this listener because only hosts can send mail while adding a comment.

Comment (HeahDude) [reply src/Form/CitizenProjectCommentCommandType.php]: See https://github.com/EnMarche/en-marche.fr/pull/1919/commits/44c680ade829c39a5b26b8a1ba9b96d5f299e912.

**Pull request merged**